workbench/src/sidebar.c

```diff
@@ -0,0 +1,1164 @@
+
+			model = gtk_tree_view_get_model(GTK_TREE_VIEW(sidebar.file_view));
+			if (gtk_tree_model_get_iter_first (model, &first))
+			{
+				sidebar_update_workbench(&first, &position);
+			}
+		}
+		break;
+	}
+}
+
+
+/* Callback function for clicking on a sidebar item */
+static void sidebar_filew_view_on_row_activated (GtkTreeView       *treeview,
+                                                 GtkTreePath       *path,
+                                                 GtkTreeViewColumn *col,
+                                                 gpointer          userdata)
+{
+	gchar *info;
+	GtkTreeModel *model;
+	GtkTreeIter  iter;
+
+	model = gtk_tree_view_get_model(treeview);
+
+	if (gtk_tree_model_get_iter(model, &iter, path))
+	{
+		guint dataid;
+		gchar *name;
+		void  *address;
+
+		gtk_tree_model_get(model, &iter, FILEVIEW_COLUMN_NAME, &name, -1);
+		gtk_tree_model_get(model, &iter, FILEVIEW_COLUMN_DATA_ID, &dataid, -1);
+		gtk_tree_model_get(model, &iter, FILEVIEW_COLUMN_ASSIGNED_DATA_POINTER, &address, -1);
+
+		switch (dataid)
+		{
+			case DATA_ID_PROJECT:
+				info = wb_project_get_info((WB_PROJECT *)address);
+				if (workbench_get_project_status_by_address(wb_globals.opened_wb, address)
+					==
+					PROJECT_ENTRY_STATUS_OK)
+				{
+					dialogs_show_msgbox(GTK_MESSAGE_INFO, "%s", info);
+				}
+				else
+				{
+					dialogs_show_msgbox(GTK_MESSAGE_ERROR, _("%s\nProject file not found!"), info);
+				}
+				g_free(info);
+			break;
+			case DATA_ID_DIRECTORY:
+			    info = wb_project_dir_get_info((WB_PROJECT_DIR *)address);
+				dialogs_show_msgbox(GTK_MESSAGE_INFO, "%s", info);
+			break;
+			case DATA_ID_WB_BOOKMARK:
+			case DATA_ID_PRJ_BOOKMARK:
+			case DATA_ID_FILE:
+				document_open_file((char *)address, FALSE, NULL, NULL);
+			break;
+			case DATA_ID_NO_DIRS:
+				dialogs_show_msgbox(GTK_MESSAGE_INFO, _("This project has no directories. Directories can be added to a project using the context menu."));
+			break;
+			default:
+			break;
+		}
+
+		g_free(name);
+	}
+}
+
+
+/* Callbac function for button release, used for the popup menu */
+static gboolean sidebar_file_view_on_button_release(G_GNUC_UNUSED GtkWidget * widget, GdkEventButton * event,
+		G_GNUC_UNUSED gpointer user_data)
+{
+	if (event->button == 3)
+	{
+		SIDEBAR_CONTEXT context;
+		POPUP_CONTEXT popup_context = POPUP_CONTEXT_BACKGROUND;
+
+		if (sidebar_file_view_get_selected_context(&context))
+		{
+			if (context.wb_bookmark != NULL)
+			{
+				popup_context = POPUP_CONTEXT_WB_BOOKMARK;
+			}
+			if (context.project != NULL)
+			{
+				popup_context = POPUP_CONTEXT_PROJECT;
+			}
+			if (context.prj_bookmark != NULL)
+			{
+				popup_context = POPUP_CONTEXT_PRJ_BOOKMARK;
+			}
+			if (context.directory != NULL)
+			{
+				popup_context = POPUP_CONTEXT_DIRECTORY;
+			}
+			if (context.folder != NULL)
+			{
+				popup_context = POPUP_CONTEXT_FOLDER;
+			}
+			if (context.file != NULL)
+			{
+				popup_context = POPUP_CONTEXT_FILE;
+			}
```

Comment:
You are right, "POPUP_CONTEXT_FILE" is the most specific situation and "POPUP_CONTEXT_WB_BOOKMARK" the least specific. Changed it.